pkg/agent/apiserver/handlers/fqdncache/handler_test.go

```diff
@@ -0,0 +1,152 @@
+	tests := []struct {
+		name                 string
+		filteredCacheEntries []types.DnsCacheEntry
+		expectedResponse     []apis.FQDNCacheResponse
+	}{
+		{
+			name: "FQDN cache exists - multiple addresses multiple domains",
+			filteredCacheEntries: []types.DnsCacheEntry{
+				{
+					FQDNName:       "example.com",
+					IPAddress:      net.ParseIP("10.0.0.1"),
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "foo.com",
+					IPAddress:      net.ParseIP("10.0.0.4"),
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "bar.com",
+					IPAddress:      net.ParseIP("10.0.0.5"),
+					ExpirationTime: expirationTime,
+				},
+			},
+			expectedResponse: []apis.FQDNCacheResponse{
+				{
+					FQDNName:       "example.com",
+					IPAddress:      "10.0.0.1",
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "foo.com",
+					IPAddress:      "10.0.0.4",
+					ExpirationTime: expirationTime,
+				},
+				{
+					FQDNName:       "bar.com",
+					IPAddress:      "10.0.0.5",
+					ExpirationTime: expirationTime,
+				},
+			},
+		},
+		{
+			name:                 "FQDN cache does not exist",
+			filteredCacheEntries: []types.DnsCacheEntry{},
+			expectedResponse:     []apis.FQDNCacheResponse{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			q := queriertest.NewMockAgentNetworkPolicyInfoQuerier(ctrl)
+			q.EXPECT().GetFQDNCache(nil).Return(tt.filteredCacheEntries)
+			handler := HandleFunc(q)
+			req, err := http.NewRequest(http.MethodGet, "", nil)
+			require.NoError(t, err)
+			recorder := httptest.NewRecorder()
+			handler.ServeHTTP(recorder, req)
+			var receivedResponse []apis.FQDNCacheResponse
+			err = json.Unmarshal(recorder.Body.Bytes(), &receivedResponse)
+			require.NoError(t, err)
+			assert.Equal(t, tt.expectedResponse, receivedResponse)
+		})
+	}
+}
+
+func TestNewFilterFromURLQuery(t *testing.T) {
+	tests := []struct {
+		name           string
+		queryParams    url.Values
+		expectedFilter *querier.FQDNCacheFilter
+		expectedError  error
+	}{
+		{
+			name:           "Empty query",
+			queryParams:    url.Values{},
+			expectedFilter: nil,
+			expectedError:  nil,
+		},
+		{
+			name: "Valid regex domain",
+			queryParams: url.Values{
+				"domain": {"example.com"},
+			},
+			expectedFilter: &querier.FQDNCacheFilter{DomainRegex: regexp.MustCompile("^example[.]com$")},
+			expectedError:  nil,
+		},
+		{
+			name: "Valid regex domain",
+			queryParams: url.Values{
+				"domain": {"*.example.com"},
+			},
+			expectedFilter: &querier.FQDNCacheFilter{DomainRegex: regexp.MustCompile("^.*[.]example[.]com$")},
+			expectedError:  nil,
+		},
+		{
+			name: "Invalid regex domain",
+			queryParams: url.Values{
+				"domain": {"^example(abc$"},
+			},
+			expectedFilter: nil,
+			expectedError:  &syntax.Error{Code: "missing closing )", Expr: "^^example(abc$$"},
```

Comment:
it's usually better to make `expectedError` a string and use `assert.ErrorContains`. This way you don't really need to know the exact error, and in this case you won't need to import `regexp/syntax`.